File: matiasmasso/Mat.NET/DT2/DTOBancTransferPool.vb
Public Class DTOBancTransferPool
    Inherits DTOBaseGuid
    Property User As DTOUser
    Property BancEmissor As DTOBanc
    Property Fch As Date
    Property Ref As String
    Property Expenses As DTOAmt
    Property Beneficiaris As List(Of DTOBancTransferBeneficiari)
    Property Cca As DTOCca
    Property Pnds As List(Of DTOPnd)

    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Shared Function Factory(oUser As DTOUser, Optional DtFch As Date = Nothing, Optional oBancEmissor As DTOBanc = Nothing, Optional oExpenses As DTOAmt = Nothing) As DTOBancTransferPool
        Dim retval As New DTOBancTransferPool
        With retval
            .User = oUser
            .Fch = If(DtFch = Nothing, DateTime.Today, DtFch)
            .BancEmissor = oBancEmissor
            .Expenses = oExpenses
            .Beneficiaris = New List(Of DTOBancTransferBeneficiari)
        End With
        Return retval
    End Function

    Public Function Total() As DTOAmt
        Dim retval = DTOAmt.Empty
        For Each oBeneficiari In _Beneficiaris
            retval.Add(oBeneficiari.Amt)
        Next
        Return retval
    End Function

    Public Function MsgId() As String
        Dim retval As String = String.Format("{0}.{1:yyyyMMddhhmmss}", _Cca.formattedId(), DateTime.Now)
        Return retval
    End Function

    Public Function DefaultFilename(Optional sBeneficiariNom As String = "") As String
        Dim sufix As String = ""
        If sBeneficiariNom > "" Then
            sufix = sBeneficiariNom
        Else
            If _Beneficiaris.Count = 1 Then
                sufix = _Beneficiaris.First.Contact.Nom
            Else
                Dim firstConcepte As String = _Beneficiaris.First.Concepte
                Dim sameConcepte As Boolean = Not _Beneficiaris.Exists(Function(x) x.Concepte <> firstConcepte)
                If sameConcepte Then
                    sufix = firstConcepte
                End If
            End If
        End If
        Dim retval As String = String.Format("SepaCreditTransfer {0} {1}.xml", _Cca.formattedId(), sufix)
        Return retval
    End Function

    Shared Function AddBeneficiari(oBancTransferPool As DTOBancTransferPool, oCta As DTOPgcCta, oContact As DTOContact, oBankBranch As DTOBankBranch, sAccount As String, oAmt As DTOAmt, sConcepte As String)
        'Quan no hi ha Partida pendent a saldar DTOPnd
        Dim retval As New DTOBancTransferBeneficiari
        With retval
            .Parent = oBancTransferPool
            .Cta = oCta
            .Contact = oContact
            .BankBranch = oBankBranch
            .Account = sAccount
            .Amt = oAmt
            .Concepte = sConcepte
        End With
        oBancTransferPool.Beneficiaris.Add(retval)
        Return retval
    End Function

    Shared Function AddBeneficiari(oBancTransferPool As DTOBancTransferPool, oCta As DTOPgcCta, oContact As DTOContact, oIban As DTOIban, oAmt As DTOAmt, sConcepte As String)
        Dim oBankBranch As DTOBankBranch = Nothing
        Dim sAccount As String = ""
        If oIban IsNot Nothing Then
            oBankBranch = oIban.BankBranch
            sAccount = oIban.Digits
        End If
        Dim retval As New DTOBancTransferBeneficiari
        With retval
            .Parent = oBancTransferPool
            .Cta = oCta
            .Contact = oContact
            .BankBranch = oBankBranch
            .Account = sAccount
            .Amt = oAmt
            .Concepte = sConcepte
        End With
        oBancTransferPool.Beneficiaris.Add(retval)
        Return retval
    End Function

    Shared Sub AddPnd(ByRef oBancTransferPool As DTOBancTransferPool, oPnd As DTOPnd, oBankBranch As DTOBankBranch, sAccount As String, Optional oLang As DTOLang = Nothing)
        If oLang Is Nothing Then oLang = DTOLang.ESP

        Dim oBeneficiari As DTOBancTransferBeneficiari = oBancTransferPool.Beneficiaris.Find(
            Function(x) x.Contact.Equals(oPnd.Contact) And x.Account = sAccount)

        If oBeneficiari Is Nothing Then
            oBeneficiari = New DTOBancTransferBeneficiari
            With oBeneficiari
                .Parent = oBancTransferPool
                .Cta = oPnd.Cta
                .Contact = oPnd.Contact
                .BankBranch = oBankBranch
                .Account = sAccount
                .Amt = DTOAmt.Empty
            End With
            oBancTransferPool.Beneficiaris.Add(oBeneficiari)
        End If

        With oBeneficiari
            .Pnds.Add(oPnd)
            .Concepte = DTOPnd.Concepte(.Pnds)
            Select Case oPnd.Cod
                Case DTOPnd.Codis.Creditor
                    .Amt.Add(oPnd.Amt)
                Case DTOPnd.Codis.Deutor
                    .Amt.Substract(oPnd.Amt)
            End Select
        End With
    End Sub

End Class

Public Class DTOBancTransferBeneficiari
    Inherits DTOBaseGuid

    Property Parent As DTOBancTransferPool
    Property Cta As DTOPgcCta
    Property Contact As DTOContact
    Property Amt As DTOAmt
    Property BankBranch As DTOBankBranch
    Property Account As String
    Property Concepte As String
    Property Pnds As List(Of DTOPnd)

    Property IsOurBankAccount As Boolean

    Public Sub New()
        MyBase.New()
        _Pnds = New List(Of DTOPnd)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _Pnds = New List(Of DTOPnd)
    End Sub

    Shared Function Ccbs(value As DTOBancTransferBeneficiari) As List(Of DTOCcb)
        'per Remmitance advice
        Dim oCca As DTOCca = value.Parent.Cca
        Dim retval As List(Of DTOCcb) = oCca.Items.FindAll(Function(x) value.Contact.Equals(x.Contact))
        Return retval
    End Function
End Class

File: matiasmasso/Mat.NET/DTO/DTOBancTransferPool.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOBancTransferPool : DTOBaseGuid
    {
        public DTOUser User { get; set; }
        public DTOBanc BancEmissor { get; set; }
        public DateTime Fch { get; set; }
        public string Ref { get; set; }
        public DTOAmt Expenses { get; set; }
        public List<DTOBancTransferBeneficiari> Beneficiaris { get; set; }
        public DTOCca Cca { get; set; }
        public List<DTOPnd> Pnds { get; set; }

        public DTOBancTransferPool() : base()
        {
        }

        public DTOBancTransferPool(Guid oGuid) : base(oGuid)
        {
        }

        public static DTOBancTransferPool Factory(DTOUser oUser, DateTime DtFch = default(DateTime), DTOBanc oBancEmissor = null/* TODO Change to default(_) if this is not a reference type */, DTOAmt oExpenses = null/* TODO Change to default(_) if this is not a reference type */)
        {
            DTOBancTransferPool retval = new DTOBancTransferPool();
            {
                var withBlock = retval;
                withBlock.User = oUser;
                withBlock.Fch = DtFch == default(DateTime) ? DTO.GlobalVariables.Today() : DtFch;
                withBlock.BancEmissor = oBancEmissor;
                withBlock.Expenses = oExpenses;
                withBlock.Beneficiaris = new List<DTOBancTransferBeneficiari>();
            }
            return retval;
        }

        public DTOAmt Total()
        {
            var retval = DTOAmt.Empty();
            foreach (var oBeneficiari in Beneficiaris)
                retval.Add(oBeneficiari.Amt);
            return retval;
        }

        public string MsgId()
        {
            string retval = string.Format("{0}.{1:yyyyMMddhhmmss}", Cca.formattedId(), DTO.GlobalVariables.Now());
            return retval;
        }

        public string DefaultFilename(string sBeneficiariNom = "")
        {
            string sufix = "";
            if (!string.IsNullOrEmpty(sBeneficiariNom))
                sufix = sBeneficiariNom;
            else if (Beneficiaris.Count == 1)
                sufix = Beneficiaris.First().Contact.Nom;
            else
            {
                string firstConcepte = Beneficiaris.First().Concepte;
                bool sameConcepte = !Beneficiaris.Exists(x => x.Concepte != firstConcepte);
                if (sameConcepte)
                    sufix = firstConcepte;
            }
            string retval = string.Format("SepaCreditTransfer {0} {1}.xml", Cca.formattedId(), sufix);
            return retval;
        }

        public static void AddBeneficiari(DTOBancTransferPool oBancTransferPool, DTOPgcCta oCta, DTOContact oContact, DTOBankBranch oBankBranch, string sAccount, DTOAmt oAmt, string sConcepte)
        {
            // Quan no hi ha Partida pendent a saldar DTOPnd
            DTOBancTransferBeneficiari retval = new DTOBancTransferBeneficiari();
            {
                var withBlock = retval;
                withBlock.Parent = oBancTransferPool;
                withBlock.Cta = oCta;
                withBlock.Contact = oContact;
                withBlock.BankBranch = oBankBranch;
                withBlock.Account = sAccount;
                withBlock.Amt = oAmt;
                withBlock.Concepte = sConcepte;
            }
            oBancTransferPool.Beneficiaris.Add(retval);
        }

        public static void AddBeneficiari(DTOBancTransferPool oBancTransferPool, DTOPgcCta oCta, DTOContact oContact, DTOIban oIban, DTOAmt oAmt, string sConcepte)
        {
            DTOBankBranch oBankBranch = null/* TODO Change to default(_) if this is not a reference type */;
            string sAccount = "";
            if (oIban != null)
            {
                oBankBranch = oIban.BankBranch;
                sAccount = oIban.Digits;
            }
            DTOBancTransferBeneficiari retval = new DTOBancTransferBeneficiari();
            {
                var withBlock = retval;
                withBlock.Parent = oBancTransferPool;
                withBlock.Cta = oCta;
                withBlock.Contact = oContact;
                withBlock.BankBranch = oBankBranch;
                withBlock.Account = sAccount;
                withBlock.Amt = oAmt;
                withBlock.Concepte = sConcepte;
            }
            oBancTransferPool.Beneficiaris.Add(retval);
        }

        public static void AddPnd(ref DTOBancTransferPool oBancTransferPool, DTOPnd oPnd, DTOBankBranch oBankBranch, string sAccount, DTOLang oLang = null/* TODO Change to default(_) if this is not a reference type */)
        {
            if (oLang == null)
                oLang = DTOLang.ESP();

            DTOBancTransferBeneficiari oBeneficiari = oBancTransferPool.Beneficiaris.Find(x => x.Contact.Equals(oPnd.Contact) & x.Account == sAccount);

            if (oBeneficiari == null)
            {
                oBeneficiari = new DTOBancTransferBeneficiari();
                {
                    var withBlock = oBeneficiari;
                    withBlock.Parent = oBancTransferPool;
                    withBlock.Cta = oPnd.Cta;
                    withBlock.Contact = oPnd.Contact;
                    withBlock.BankBranch = oBankBranch;
                    withBlock.Account = sAccount;
                    withBlock.Amt = DTOAmt.Empty();
                }
                oBancTransferPool.Beneficiaris.Add(oBeneficiari);
            }

            {
                var withBlock = oBeneficiari;
                withBlock.Pnds.Add(oPnd);
                withBlock.Concepte = DTOPnd.Concepte(withBlock.Pnds);
                switch (oPnd.Cod)
                {
                    case DTOPnd.Codis.Creditor:
                        {
                            withBlock.Amt.Add(oPnd.Amt);
                            break;
                        }

                    case DTOPnd.Codis.Deutor:
                        {
                            withBlock.Amt.Substract(oPnd.Amt);
                            break;
                        }
                }
            }
        }
    }

    public class DTOBancTransferBeneficiari : DTOBaseGuid
    {
        public DTOBancTransferPool Parent { get; set; }
        public DTOPgcCta Cta { get; set; }
        public DTOContact Contact { get; set; }
        public DTOAmt Amt { get; set; }
        public DTOBankBranch BankBranch { get; set; }
        public string Account { get; set; }
        public string Concepte { get; set; }
        public List<DTOPnd> Pnds { get; set; }

        public bool IsOurBankAccount { get; set; }

        public DTOBancTransferBeneficiari() : base()
        {
            Pnds = new List<DTOPnd>();
        }

        public DTOBancTransferBeneficiari(Guid oGuid) : base(oGuid)
        {
            Pnds = new List<DTOPnd>();
        }

        public static List<DTOCcb> Ccbs(DTOBancTransferBeneficiari value)
        {
            // per Remmitance advice
            DTOCca oCca = value.Parent.Cca;
            List<DTOCcb> retval = oCca.Items.FindAll(x => value.Contact.Equals(x.Contact));
            return retval;
        }
    }
}
